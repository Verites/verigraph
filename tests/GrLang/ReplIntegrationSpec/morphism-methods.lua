GrLang.add_node_type('N')
GrLang.add_edge_type('E', 'N', 'N')

function with_dom(...)
  if select('#', ...) > 0 then
    local fst = ...
    return fst:dom(), ...
  else
    return ...
  end
end

function with_cod(...)
  if select('#', ...) > 0 then
    local fst = ...
    return fst:cod(), ...
  else
    return ...
  end
end

X = Graph [[ a b c : N ]] 
Y = Graph [[ e f g h i : N ]]

f = Morphism(X, Y)[[a->e; b->f; c->g]]
print('dom', f:dom() == X)
print('cod', f:cod() == Y)
print('is_monic,epic,iso', f:is_monic(), f:is_epic(), f:is_iso())

print('dot')
print(f:to_dot())
print()

Z = Graph [[ae bfg c h i: N]]
f, g = Morphism(X,Z)[[a->ae; b->bfg; c->c]], Morphism(Y,Z)[[e->ae; f g ->bfg; h->h; i->i]]
print('pullback', with_dom(f:pullback(g)))

W = Graph [[ae bf : N]]
f, g = Morphism(W,X)[[ae->a; bf->b]], Morphism(W,Y)[[ae->e; bf->f]]
print('pushout', with_cod(f:pushout(g)))

f, g = Morphism(X,Y)[[a->e; b->f; c->g]], Morphism(X,Y)[[a->e; b->f; c->h]]
print('equalize_with', with_dom(f:equalize_with(g)))
print('coequalize_with', with_cod(f:coequalize_with(g)))

f = Morphism(X, Y)[[a->e; b->f; c->h]]
g = Y:identity()
print('has_pushout_complement', f:has_pushout_complement(g))
print('pushout_complement', with_dom(f:pushout_complement(g)))

g = Morphism(Y, Y)[[e->e; f->f; g h->g; i->i]]
print('has_pushout_complement', f:has_pushout_complement(g))
print('pushout_complement', with_dom(f:pushout_complement(g)))

g = Morphism(Y, Y)[[e->e; f->f; g ->g; h i->h]]
b,gg,c = g:initial_pushout()
print('initial_pushout', b:dom(), b, c:dom(), c, gg)

W = Graph [[ac : N]]
X = Graph [[a b: N]]
Y = Graph [[c d : N]]
f, g = Morphism(W,X)[[ac->a]], Morphism(W,Y)[[ac->c]]
print()
print('commutative_overlappings_with(all)')
for h,k in f:commutative_overlappings_with(g) do
  print(with_cod(h, k))
end

print()
print('commutative_overlappings_with(monic)')
for h,k in f:commutative_overlappings_with(g, 'monic') do
  print(with_cod(h, k))
end