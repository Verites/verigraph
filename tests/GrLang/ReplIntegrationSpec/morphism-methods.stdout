dom	true
cod	true
is_monic,epic,iso	true	false	false
pullback	graph  {
  ?0 ?1 ?2 : N
}	morphism  : dom -> cod {
  ?0 -> a
  ?2 ?1 -> b
}	morphism  : dom -> cod {
  ?0 -> e
  ?1 -> f
  ?2 -> g
}
pushout	graph  {
  a b c g h : N
  i : N
}	morphism  : dom -> cod {
  a -> a
  b -> b
  c -> c
}	morphism  : dom -> cod {
  e -> a
  f -> b
  g -> g
  h -> h
  i -> i
}
equalize_with	graph  {
  a b : N
}	morphism  : dom -> cod {
  a -> a
  b -> b
}
coequalize_with	graph  {
  e f g i : N
}	morphism  : dom -> cod {
  e -> e
  f -> f
  h g -> g
  i -> i
}
has_pushout_complement	true
pushout_complement	graph  {
  a b c : N
}	morphism  : dom -> cod {
  a -> e
  b -> f
  c -> h
}	morphism  : dom -> cod {
  e -> e
  f -> f
  h -> h
}
has_pushout_complement	false
pushout_complement
initial_pushout	graph  {
  ?3 ?4 : N
}	morphism  : dom -> cod {
  ?3 -> h
  ?4 -> i
}	graph  {
  h i : N
}	morphism  : dom -> cod {
  h -> h
  i -> i
}	morphism  : dom -> cod {
  ?4 ?3 -> h
}

commutative_overlappings_with(all)
graph  {
  ?3 ?1 ?2 : N
}	morphism  : dom -> cod {
  b -> ?1
  a -> ?2
}	morphism  : dom -> cod {
  c -> ?2
  d -> ?3
}
graph  {
  ?3 ?2 : N
}	morphism  : dom -> cod {
  a -> ?2
  b -> ?3
}	morphism  : dom -> cod {
  c -> ?2
  d -> ?3
}
graph  {
  ?1 ?3 : N
}	morphism  : dom -> cod {
  b -> ?1
  a -> ?3
}	morphism  : dom -> cod {
  d c -> ?3
}
graph  {
  ?3 ?2 : N
}	morphism  : dom -> cod {
  b a -> ?2
}	morphism  : dom -> cod {
  c -> ?2
  d -> ?3
}
graph  {
  ?3 : N
}	morphism  : dom -> cod {
  b a -> ?3
}	morphism  : dom -> cod {
  d c -> ?3
}

commutative_overlappings_with(monic)
graph  {
  ?3 ?1 ?2 : N
}	morphism  : dom -> cod {
  b -> ?1
  a -> ?2
}	morphism  : dom -> cod {
  c -> ?2
  d -> ?3
}
graph  {
  ?3 ?2 : N
}	morphism  : dom -> cod {
  a -> ?2
  b -> ?3
}	morphism  : dom -> cod {
  c -> ?2
  d -> ?3
}
